### PR TITLE
Separate controller manager test cases

### PR DESCRIPTION
**Pull request by bmagyar**
Trying to unentangle these larger collection of test cases depending on each other.

### CODE REVIEW

**Comment by bmagyar** [comment]
@destogl this is really ready now

File: controller_manager/test/test_load_controller.cpp [hunk 1]
```diff
@@ -183,192 +213,314 @@ TEST_F(TestLoadController, configure_controller)
...
-  EXPECT_EQ(
-    cm_->configure_controller(controller_name1), controller_interface::return_type::OK);
-  ASSERT_EQ(
-    lifecycle_msgs::msg::State::PRIMARY_STATE_INACTIVE,
-    abstract_test_controller1.c->get_current_state().id());
-  EXPECT_EQ(1u, cleanup_calls);
 }
 
-TEST_F(TestLoadController, switch_controller_empty)
+TEST_F(TestLoadController, inactive_controller_cannot_be_configured)
```
Comment (destogl) [inline controller_manager/test/test_load_controller.cpp:297]: This test should be moved up before `can_not_configure_active_controller`, or the `can_not_configure_active_controller` should be at least after `can_stop_active_controller`.

In that was the tests would be like: “normal lifecycle”, "errors in lifecycle"

File: controller_manager/test/test_load_controller.cpp [hunk 2]
```diff
@@ -183,192 +213,314 @@ TEST_F(TestLoadController, configure_controller)
...
+    std::dynamic_pointer_cast<test_controller::TestController>(controller_if);
+  size_t cleanup_calls = 0;
+  test_controller->cleanup_calls = &cleanup_calls;
+  // Configure from inactive state
+  test_controller->simulate_cleanup_failure = false;
+  EXPECT_EQ(cm_->configure_controller(controller_name1), controller_interface::return_type::OK);
+  ASSERT_EQ(
+    lifecycle_msgs::msg::State::PRIMARY_STATE_INACTIVE,
+    controller_if->get_current_state().id());
+  EXPECT_EQ(1u, cleanup_calls);
```
Comment (destogl) [inline controller_manager/test/test_load_controller.cpp:353]: Why is here cleanup called?

File: controller_manager/test/test_load_controller.cpp [hunk 3]
```diff
@@ -183,192 +213,314 @@ TEST_F(TestLoadController, configure_controller)
...
+    std::make_tuple(
+      controller_interface::return_type::OK, BEST_EFFORT, NONEXISTENT_CONTROLLER,
+      NONEXISTENT_CONTROLLER, "Switch with nonexistent start and stop controllers specified"),
+    // valid controller used
+    std::make_tuple(
+      controller_interface::return_type::ERROR, STRICT, NONEXISTENT_CONTROLLER,
+      VALID_CONTROLLER, "Switch with valid stopped controller specified"),
+    std::make_tuple(
+      controller_interface::return_type::OK, BEST_EFFORT, NONEXISTENT_CONTROLLER,
+      VALID_CONTROLLER, "Switch with valid stopped controller specified"),
```
Comment (destogl) [inline controller_manager/test/test_load_controller.cpp:443]: Error messages do not seem to be right, at least they are not clear. What about to make them longer and more descriptive?

File: controller_manager/test/test_load_controller.cpp [hunk 4]
```diff
@@ -183,192 +213,314 @@ TEST_F(TestLoadController, configure_controller)
...
+      "Switch with no controllers specified"),
+    std::make_tuple(
+      controller_interface::return_type::OK, BEST_EFFORT, EMPTY_STR_VEC,
+      EMPTY_STR_VEC, "Switch with no controllers specified"),
+    // combination of empty and non-existent controller
+    std::make_tuple(
+      controller_interface::return_type::OK, UNSPECIFIED, NONEXISTENT_CONTROLLER,
+      EMPTY_STR_VEC, "Switch with nonexistent controller specified"),
+    std::make_tuple(
+      controller_interface::return_type::ERROR, STRICT, NONEXISTENT_CONTROLLER,
```
Comment (destogl) [inline controller_manager/test/test_load_controller.cpp:472]: probably the same comment about error messages as above.

File: controller_manager/test/test_load_controller.cpp [hunk 5]
```diff
@@ -183,192 +213,314 @@ TEST_F(TestLoadController, configure_controller)
...
+  ASSERT_EQ(
+    lifecycle_msgs::msg::State::PRIMARY_STATE_INACTIVE,
+    controller_if->get_current_state().id());
+  EXPECT_EQ(1u, cleanup_calls);
 }
 
-TEST_F(TestLoadController, switch_controller)
-{
-  std::string controller_type = test_controller::TEST_CONTROLLER_CLASS_NAME;
+class SwitchTest : public TestLoadController,
```
Comment (destogl) [inline controller_manager/test/test_load_controller.cpp:356]:
```suggestion
class SwitchErrorsEmptyNonexistentTest : public TestLoadController,
```

File: controller_manager/test/test_load_controller.cpp [hunk 6]
```diff
@@ -183,192 +213,314 @@ TEST_F(TestLoadController, configure_controller)
...
+    std::make_tuple(
+      controller_interface::return_type::ERROR, STRICT, VALID_PLUS_NONEXISTENT_CONTROLLERS,
+      EMPTY_STR_VEC, "Switch with valid and nonexistent controller specified"),
+    std::make_tuple(
+      controller_interface::return_type::ERROR, STRICT, VALID_CONTROLLER,
+      NONEXISTENT_CONTROLLER, "Switch with  valid and nonexistent controller specified")
+));
+#endif
+
+TEST_F(TestLoadController, starting_and_stopping_a_controller)
```
Comment (destogl) [inline controller_manager/test/test_load_controller.cpp:511]: Isn't this test done already above? I think it was "test_configure_inactive_controller".

**Review by destogl — COMMENTED** [review]
A few comments to optimize this further.

**Comment by bmagyar** [comment]
@destogl created a good first issue capturing your points here: https://github.com/ros-controls/ros2_control/issues/493

**Review by destogl — APPROVED** [review]
Ok, comment will be addressed in follow-ups.

**Pull request merged**